File: PARQme/src/com/objects/SavedInfo.java
package com.objects;

import java.util.Date;

import android.content.Context;
import android.content.SharedPreferences;

public class SavedInfo{
	/* Constantly making the same check and editor and changing info clutters code, dififcult to see logic.  */
	/* parkState
	 * String email
	 * String endTime
	 * String code
	 * Boolean remember
	 * */
	//TODO:  make sure all needed info is set via reset and stuff.
	public static final String SAVED_INFO = "ParqMeInfo";
	public static boolean isFirstTime(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		return check.getBoolean("firstTimeFlag", true);
	}
	public static boolean ringEnable(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		return check.getBoolean("ringEnable", true);
	}
	public static boolean warningEnable(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		return check.getBoolean("warningEnable", true);
	}
	public static boolean vibrateEnable(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		return check.getBoolean("vibrateEnable", true);
	}
	public static boolean autoRefill(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		return check.getBoolean("autoRefill", true);
	}
	public static void eraseTimer(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		SharedPreferences.Editor editor = check.edit();
		editor.putLong("endTime", 0);
		editor.commit();
	}
	//if parked return true, else false
	public static boolean isParked(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		return check.getLong("endTime", 0) > System.currentTimeMillis();
	}
	//if logged in returns true, else false
	public static boolean isLoggedIn(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO,0);
		return check.getLong("uid", -1) != -1;
		//SharedPreferences.Editor editor = check.edit();
	}
	//set's email to a string
	public static void setEmail(Context activity, String email){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		editor.putString("email", email);
		editor.commit();
		
	}
	public static void setLatLon(Context activity, double lat, double lon){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		editor.putString("lat", ""+lat);
		editor.putString("lon", ""+lon);
		editor.commit();
		
	}
	public static void setGeneric(Context activity, String title, boolean input){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		editor.putBoolean(title, input);
		editor.commit();
	}
	public static long getEndTime(SharedPreferences prefs){
		return prefs.getLong("endTime", 0);
	}
	public static Date getEndTimeDate(SharedPreferences prefs) {
		return new Date(getEndTime(prefs));
	}
	public static String getEmail(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		return check.getString("email", null);
	}
	public static String getCardStub(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		return check.getString("ccStub", "****");
	}
	public static String getLat(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		return check.getString("lat", "0");
	}
	public static String getLon(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		return check.getString("lon", "0");
	}
	//set's code to a string
	public static void setCode(Context activity, String code){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		editor.putString("code", code);
		editor.commit();
		
	}
	public static void setParkingReferenceNumber(Context activity, String parkReferenceNumber){

		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		editor.putString("PARKID", parkReferenceNumber);
		editor.commit();
	}
	public static String getParkId(SharedPreferences prefs) {
	    return prefs.getString("PARKID", "");
	}
    public static void unpark(Context activity) {
    	SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor edit = check.edit();
        edit.remove("PARKID");
        edit.remove("endTime");
        edit.remove("lat");
        edit.remove("lon");
        edit.remove("spot");
        edit.remove("minTime");
        edit.remove("maxTime");
        edit.remove("defaultRate");
        edit.remove("minIncrement");
        edit.remove("description");
        edit.commit();
    }
    public static void park(Context activity, ParkInstanceObject park, RateObject rate) {
        SharedPreferences.Editor edit = activity.getSharedPreferences(SAVED_INFO, 0).edit();
        edit.putString("PARKID", park.getParkingReferenceNumber());
        edit.putLong("endTime", park.getEndTime());
        edit.putString("lat", String.valueOf(rate.getLat()));
        edit.putString("lon", String.valueOf(rate.getLon()));
        edit.putLong("spot", rate.getSpot());
        edit.putInt("minTime", rate.getMinTime());
        edit.putInt("maxTime", rate.getMaxTime());
        edit.putInt("defaultRate", rate.getDefaultRate());
        edit.putInt("minIncrement", rate.getMinIncrement());
        edit.putString("description", rate.getDescription());
        edit.commit();
    }
    public static RateObject getRate(SharedPreferences prefs) {
        final double lat = Double.valueOf(prefs.getString("lat", "0"));
        final double lon = Double.valueOf(prefs.getString("lon", "0"));
        final long spot = prefs.getLong("spot", 0);
        final int minTime = prefs.getInt("minTime", 0);
        final int maxTime = prefs.getInt("maxTime", 0);
        final int defaultRate = prefs.getInt("defaultRate", 0);
        final int minIncrement = prefs.getInt("minIncrement", 0);
        final String description = prefs.getString("description", "");
        return new RateObject(lat, lon, spot, minTime, maxTime, defaultRate, minIncrement, description);
    }
	public static void logIn(Context activity, boolean parkState, String email, long uid, String password, String ccStub){
		SharedPreferences.Editor editor = activity.getSharedPreferences(SAVED_INFO, 0).edit();
		editor.putString("email", email);
		editor.putLong("uid", uid);
		editor.putString("password", password);
		editor.putString("ccStub", ccStub);
		editor.commit();
	}
    public static void logOut(Context activity){
        SharedPreferences.Editor editor = activity.getSharedPreferences(SAVED_INFO, 0).edit();
        editor.clear();
        editor.commit();
    }
	
	public static void syncParkingSession(Context activity, ParkSync sync){
		SharedPreferences.Editor editor = activity.getSharedPreferences(SAVED_INFO, 0).edit();
		editor.putInt("defaultRate", sync.getDefaultRate());
		editor.putString("description",sync.getDescription());
		editor.putLong("endTime", sync.getEndTime());
		editor.putString("lat", ""+sync.getLat());
		editor.putString("lon", ""+sync.getLon());
		editor.putInt("maxTime", sync.getMaxTime());
		editor.putInt("minIncrement", sync.getMinIncrement());
		editor.putInt("minTime", sync.getMinTime());
		editor.putString("PARKID", sync.getParkingReferenceNumber());
		editor.putLong("spot", sync.getSpotId());
		editor.commit();
	}
	public static void toggleVibrate(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		if(check.getBoolean("vibrateEnable", false)){
			//if parked, set as not parked
			editor.putBoolean("vibrateEnable", false);
		}else{
			//if not parked, set as parked.
			editor.putBoolean("vibrateEnable", true);
		}
		editor.commit();
	}
	public static void toggleWarn(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		if(check.getBoolean("warningEnable", false)){
			//if parked, set as not parked
			editor.putBoolean("warningEnable", false);
		}else{
			//if not parked, set as parked.
			editor.putBoolean("warningEnable", true);
		}
		editor.commit();
	}
	public static void toggleRing(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		if(check.getBoolean("ringEnable", false)){
			//if parked, set as not parked
			editor.putBoolean("ringEnable", false);
		}else{
			//if not parked, set as parked.
			editor.putBoolean("ringEnable", true);
		}
		editor.commit();
	}
	public static void toggleRefill(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		if(check.getBoolean("autoRefill", false)){
			//if parked, set as not parked
			editor.putBoolean("autoRefill", false);
		}else{
			//if not parked, set as parked.
			editor.putBoolean("autoRefill", true);
		}
		editor.commit();
	}
	public static void toggleRemember(Context activity){
		SharedPreferences check = activity.getSharedPreferences(SAVED_INFO, 0);
		SharedPreferences.Editor editor = check.edit();
		if(check.getBoolean("remember", false)){
			//if parked, set as not parked
			editor.putBoolean("remember", false);
		}else{
			//if not parked, set as parked.
			editor.putBoolean("remember", true);
		}
		editor.commit();
	}
}

File: parq-ios/Parq/SavedInfo.m
//
//  SavedInfo.m
//  hello
//
//  Created by Michael Xia on 1/23/12.
//  Copyright (c) 2012 Iudex Projects. All rights reserved.
//

#import "SavedInfo.h"

@implementation SavedInfo
+(NSNumber*) getUid{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"uid"];
}
+(NSString*) getPlistPath{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSError *error;
    NSString *documentsDirectory = [paths objectAtIndex:0];
    NSString *path = [documentsDirectory stringByAppendingPathComponent:@"ParqInfo.plist"];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath: path]){
        //if the file doesn't exist yet, create it and do write.  
        NSString *bundle = [[NSBundle mainBundle] pathForResource:@"ParqInfo" ofType:@"plist"];
        [fileManager copyItemAtPath:bundle toPath: path error:&error]; 
    }
    return path;
}

+(BOOL) ringEnable{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [[savedStock objectForKey:@"ringEnable"] boolValue];
}
+(BOOL) vibrateEnable{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [[savedStock objectForKey:@"vibrateEnable"] boolValue];
}
+(BOOL) autoRefill{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [[savedStock objectForKey:@"autoRefill"] boolValue];
}
+(BOOL) isParked{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    NSNumber* endTime = [savedStock objectForKey:@"endTime"];
    return [endTime doubleValue]>CACurrentMediaTime();
}
+(BOOL) isLoggedIn{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    NSNumber* result = [savedStock objectForKey:@"uid"];
    if(result ==nil) return NO;
    return [result intValue]!=-1;
}
+(void) setEmail:(NSString*)emailIn{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:emailIn forKey:@"email"];        
    [data writeToFile: path atomically:YES];
}
+(void) setLat:(NSNumber*)latIn Lon:(NSNumber*)lonIn{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:latIn forKey:@"lat"];
    [data setObject:lonIn forKey:@"lon"];
    [data writeToFile: path atomically:YES];
}
+(void) setGeneric:(NSString*) title boolInput:(BOOL)mBoolean{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:[NSNumber numberWithBool:mBoolean ] forKey:title];        
    [data writeToFile: path atomically:YES];
}
+(NSNumber*) getEndTime{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"endTime"];
}
+(NSString*) getEmail{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"email"];
}
+(NSString*) getCardStub{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"ccStub"];
}
+(NSNumber*) getLat{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"lat"];
}
+(NSNumber*) getLon{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"lon"];
}
+(void) setParkingReferenceNumber:(NSString*) parkRefNumIn{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:parkRefNumIn forKey:@"parkRef"];        
    [data writeToFile: path atomically:YES];
}
+(NSString*) getParkRefNum{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"parkRef"];
}
+(void) unpark{
    NSArray* keys = [NSArray arrayWithObjects:@"endTime", @"lat", @"lon", @"spot", @"minTime", @"maxTime", @"defaultRate", @"minIncrement", @"description", @"spotNumber", @"code", nil];
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data removeObjectsForKeys:keys];
    [data writeToFile:path atomically:YES];
}
+(void) park:(ParkInstanceObject*)parkInstObjIn Rate:(RateObject*) rateObjIn{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:parkInstObjIn.parkingReferenceNumber forKey:@"parkRef"];
    [data setObject:parkInstObjIn.endTime forKey:@"endTime"];
    [data setObject:rateObjIn.spotNumber forKey:@"spot"];
    [data setObject:rateObjIn.minTime forKey:@"minTime"];
    [data setObject:rateObjIn.rateCents forKey:@"defaultRate"];
    [data setObject:rateObjIn.minuteInterval forKey:@"minIncrement"];
    [data setObject:rateObjIn.description forKey:@"description"];
    [data writeToFile: path atomically:YES];
}
+(RateObject*) getRate{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    return [[RateObject alloc] initWithLat:[data objectForKey:@"lat"] lon:[data objectForKey:@"lon"] spot:[data objectForKey:@"spot"] min:[data objectForKey:@"minTime"] max:[data objectForKey:@"maxTime"] defRate:[data objectForKey:@"defaultRate"] minInc:[data objectForKey:@"minIncrement"] desc:[data objectForKey:@"description"]];
}
+(void) logIn:(NSNumber*) parkState Email:(NSString*)emailIn UID:(NSNumber*) userId ccStub:(NSString*)ccStubIn{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:parkState forKey:@"parkState"];        
    [data setObject:emailIn forKey:@"email"];        
    [data setObject:userId forKey:@"uid"];        
    [data setObject:ccStubIn forKey:@"ccStub"];        
    [data writeToFile: path atomically:YES];
}
+(void) logOut{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data removeAllObjects];
    [data writeToFile:path atomically:YES];
}
+(void) syncParkingSession:(ParkSync*)sync{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:sync.defaultRate forKey:@"defaultRate"];        
    [data setObject:sync.desc forKey:@"description"];        
    [data setObject:sync.endTime forKey:@"endTime"];        
    [data setObject:sync.lat forKey:@"lat"];           
    [data setObject:sync.lon forKey:@"lon"];       
    [data setObject:sync.maxTime forKey:@"maxTime"];       
    [data setObject:sync.minTime forKey:@"minTime"];       
    [data setObject:sync.minInc forKey:@"minIncrement"];       
    [data setObject:sync.parkingReferenceNumber forKey:@"parkRef"];       
    [data setObject:sync.spotId forKey:@"spot"];       
    [data setObject:sync.spotNumber forKey:@"spotNumber"];    
    [data writeToFile: path atomically:YES];
}
+(NSString*) getSpotNumber{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"spotNumber"];
}
+(NSNumber*) getSpotId{
    NSMutableDictionary* savedStock = [[NSMutableDictionary alloc] initWithContentsOfFile:[self getPlistPath]];
    return [savedStock objectForKey:@"spot"];
}
+(void) setSpotNumber:(NSString*)spotNumIn{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    [data setObject:spotNumIn forKey:@"spot"];        
    [data writeToFile: path atomically:YES];
}
+(void) toggleVibrate{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    if([[data objectForKey:@"vibrateEnable"] boolValue]){
        [data setObject:[NSNumber numberWithBool:NO] forKey:@"vibrateEnable"];
    }else{
        [data setObject:[NSNumber numberWithBool:YES] forKey:@"vibrateEnable"];
    }
    [data writeToFile:path atomically:YES];
}
+(void) toggleRing{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    if([[data objectForKey:@"ringEnable"] boolValue]){
        [data setObject:[NSNumber numberWithBool:NO] forKey:@"ringEnable"];
    }else{
        [data setObject:[NSNumber numberWithBool:YES] forKey:@"ringEnable"];
    }
    [data writeToFile:path atomically:YES];
    
}
+(void) toggleRefill{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    if([[data objectForKey:@"autoRefill"] boolValue]){
        [data setObject:[NSNumber numberWithBool:NO] forKey:@"autoRefill"];
    }else{
        [data setObject:[NSNumber numberWithBool:YES] forKey:@"autoRefill"];
    }
    [data writeToFile:path atomically:YES];
}
+(void) toggleRemember{
    NSString* path = [self getPlistPath];
    NSMutableDictionary *data = [[NSMutableDictionary alloc] initWithContentsOfFile: path];
    if([[data objectForKey:@"remember"] boolValue]){
        [data setObject:[NSNumber numberWithBool:NO] forKey:@"remember"];
    }else{
        [data setObject:[NSNumber numberWithBool:YES] forKey:@"remember"];
    }
    [data writeToFile:path atomically:YES];
}

@end
